File: ruby/left_center_right.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

class Die
  def self.roll
    %i[left center right dot dot dot].sample
  end
end

class Player
  attr_reader :name
  attr_accessor :chip_count

  def initialize(name, chip_count)
    @name = name
    @chip_count = chip_count
  end

  def has_chips?
    chip_count.positive?
  end

  def transfer_chip_to(player)
    return unless (player && has_chips?)
    self.chip_count -= 1
    player.chip_count += 1
  end

  def roll
    [Die.roll, Die.roll, Die.roll].first([chip_count, 3].min)
  end
end

class LeftCenterRight
  attr_reader :players, :starting_chips
  attr_accessor :pot

  # player names are specified in order of play around the table clock-wise
  def initialize(starting_chips, *player_names)
    raise 'LCR needs at least three players!' if player_names.size < 3
    raise 'LCR needs at least one chip!' if starting_chips < 1
    @players = player_names.map do |name|
      Player.new(name, starting_chips)
    end
    @pot = Player.new('pot', 0) # a Player for ease of chip transfer
    @starting_chips = starting_chips
    @play_counter = 0
  end

  def play
    reset
    play_turn until game_over?
    puts "#{winner} wins the game in #{@play_counter} plays!"
  end

  private

  def reset
    players.each { |p| p.chip_count = starting_chips }
    @pot.chip_count = 0
    @play_counter = 0
  end

  def handle_turn_for(player)
    transfer_targets = {
      left:   players.last,
      center: pot,
      right:  players[1],
      # dot:  player      # no point in transferring to self
    }
    print "[#{@play_counter}] #{player.name} rolls:"
    player.roll.each do |roll|
      print " #{roll}"
      player.transfer_chip_to(transfer_targets[roll])
    end
    print "\n"
  end

  def play_turn
    @play_counter += 1
    active_player = players.first
    handle_turn_for(active_player) if active_player.has_chips?
    show_state
    players.rotate!
  end

  def winner
    players.each do |player|
      return player.name if player.has_chips?
    end
    nil
  end

  def game_over?
    players_with_chips = players.map(&:has_chips?).count(true)
    players_with_chips == 1
  end

  def show_state
    print '| '
    players.each do |player|
      print "#{player.name}: #{player.chip_count}, "
    end
    print "#{pot.name}: #{pot.chip_count}\n"
  end
end

if $PROGRAM_NAME == __FILE__
  LeftCenterRight.new(5, 'Alice', 'Bob', 'Charlie', 'Donna').play
end

File: javascript/left_center_right.js
class Die {
  static roll() {
    const SIDES = ['left', 'center', 'right', 'dot', 'dot', 'dot'];
    return SIDES[Math.floor(Math.random()*SIDES.length)];
  }
}

class Player {
  constructor(name, chip_count) {
    this.name = name;
    this.chip_count = chip_count;
  }

  hasChips() {
    return this.chip_count > 0;
  }

  transferChipTo(player) {
    if (!player || !this.hasChips()) return;
    this.chip_count -= 1;
    player.chip_count += 1;
  }

  roll() {
    let dice_to_roll = Math.min(this.chip_count, 3);
    let rolls = [];
    for (let i=0; i < dice_to_roll; i++) {
      rolls.push(Die.roll());
    }
    return rolls;
  }
}

class LeftCenterRight {
  constructor(starting_chips, ...player_names) {
    if (player_names.length < 3) throw 'LCR needs at least three players!';
    if (starting_chips < 1) throw 'LCR needs at least one chip!';
    this.starting_chips = starting_chips;
    this.players = player_names.map(function(name) {
      return new Player(name, starting_chips);
    });
    this.pot = new Player('pot', 0); // a Player for ease of chip transfer
    this.play_counter = 0;
    this.max_plays = 100;
  }

  play() {
    this._reset();
    do { this._play_turn(); } while (! this._game_over());
    return {
      winner: this._winner(),
      total_turns: this.play_counter
    };
  }

  // ----- private functions

  _reset() {
    let game = this;
    this.players.forEach(function(player) {
      player.chip_count = this.starting_chips;
    }.bind(game));
    this.pot.chip_count = 0;
    this.play_counter = 0;
  }

  _play_turn() {
    let active_player = this._active_player();
    if (active_player.hasChips()) {
      this._handle_turn();
      if (this.debug) { this._show_state(); }
    }
    this.play_counter += 1;
  }

  _active_player(offset = 0) {
    let player_index = (offset + this.play_counter + this.players.length) % this.players.length;
    return this.players[player_index];
  }

  _handle_turn() {
    let debug = this.debug;
    let left_player   = this._active_player(-1);
    let active_player = this._active_player(0);
    let right_player  = this._active_player(1);
    let transfer_targets = {
      left: left_player,
      center: this.pot,
      right: right_player
      // dot: active_player // no point in transferring to self though
    };
    if (debug) { process.stdout.write(`[${this.play_counter}] ${active_player.name} rolls:`); }
    active_player.roll().forEach(function(roll) {
      if (debug) { process.stdout.write(' ' + roll); }
      active_player.transferChipTo(transfer_targets[roll])
    });
    if (debug) { process.stdout.write('\n'); }
  }

  _winner() {
    let player_with_chips = this.players.find(function(player) {
      return player.hasChips();
    });
    if (player_with_chips === undefined) { return 'Nobody'; }
    return player_with_chips.name;
  }

  _game_over() {
    let players_with_chips = 0;
    this.players.forEach(function(player) {
      if (player.chip_count > 0 ) { players_with_chips += 1; }
    });
    if (this.play_counter >= this.max_plays) { return true; }
    return (players_with_chips === 1);
  }

  _show_state() {
    process.stdout.write('| ');
    this.players.forEach(function(player) {
      process.stdout.write(`${player.name}: ${player.chip_count}, `);
    });
    process.stdout.write(`${this.pot.name}: ${this.pot.chip_count}\n`);
  }
}

module.exports = { Die, Player, LeftCenterRight };
